src/main/java/com/gtnewhorizon/gtnhlib/capability/item/InventoryItemSink.java

```diff
@@ -0,0 +1,173 @@
+        if (slots == null) {
+            allowedSlots = null;
+            return;
+        }
+
+        if (allowedSlots == null) {
+            allowedSlots = new BitSet();
+        } else {
+            allowedSlots.clear();
+        }
+
+        for (int slot : slots) {
+            allowedSlots.set(slot);
+        }
+    }
+
+    @Override
+    public ItemStack store(ItemStack stack) {
+        stack = ItemStack.copyItemStack(stack);
+
+        if (stack.stackSize <= 0) return null;
+
+        this.markedDirty = false;
+
+        IntIterable slots;
+
+        if (inv instanceof ISidedInventory sided) {
+            slots = IntArrayList.wrap(sided.getAccessibleSlotsFromSide(side.ordinal()));
+        } else {
+            slots = () -> IntIterators.fromTo(0, inv.getSizeInventory());
+        }
+
+        IterableBitSet emptySlots = null;
+
+        for (IntIterator iterator = slots.iterator(); iterator.hasNext();) {
+            int slot = iterator.nextInt();
+
+            if (allowedSlots != null && !allowedSlots.get(slot)) continue;
+
+            ItemStack inSlot = inv.getStackInSlot(slot);
+
+            if (ItemUtil.isStackEmpty(inSlot)) {
+                if (emptySlots == null) emptySlots = new IterableBitSet();
+                emptySlots.set(slot);
+                continue;
+            }
+
+            if (!ItemUtil.areStacksEqual(inSlot, stack)) continue;
+
+            if (!inv.isItemValidForSlot(slot, stack)) continue;
+
+            if (side != ForgeDirection.UNKNOWN && inv instanceof ISidedInventory sided) {
+                if (!sided.canInsertItem(slot, stack, side.ordinal())) continue;
+            }
+
+            int slotLimit = getSlotStackLimit(slot, inSlot);
+            int remaining = slotLimit - inSlot.stackSize;
+
+            int toTransfer = Math.min(remaining, stack.stackSize);
+
+            stack.stackSize -= toTransfer;
+            inSlot.stackSize += toTransfer;
+
+            inv.setInventorySlotContents(slot, inSlot);
+
+            markDirty();
+
+            if (stack.stackSize <= 0) return null;
+        }
+
+        if (stack.stackSize <= 0) return null;
+        if (emptySlots == null) return stack;
+
+        for (int slot : emptySlots) {
+            if (!inv.isItemValidForSlot(slot, stack)) continue;
+
+            if (side != ForgeDirection.UNKNOWN && inv instanceof ISidedInventory sided) {
+                if (!sided.canInsertItem(slot, stack, side.ordinal())) continue;
+            }
+
+            int slotStackLimit = getSlotStackLimit(slot, stack);
+            int toTransfer = Math.min(slotStackLimit, stack.stackSize);
+
+            ItemStack inserted = ItemUtil.copyAmount(toTransfer, stack);
+            inv.setInventorySlotContents(slot, inserted);
+            stack.stackSize -= toTransfer;
+
+            markDirty();
+
+            if (stack.stackSize <= 0) return null;
+        }
+
+        return stack.stackSize <= 0 ? null : stack;
```

Comment:
Again, should be enough to just return stack, should not be possible to be 0 amount here.